File: Practise/stack_using_LL.c
#include <stdio.h>
#include<stdlib.h>

struct Node{
    int data;
    struct Node* next;
};

struct Node* top = NULL;

void push(int val){
    struct Node* newNode = (struct Node*)malloc(sizeof(struct Node));
    if(!newNode){
        printf("Heap Overflow\n");
        return;
    }
    newNode->data = val;
    newNode->next = top;
    top = newNode;
    printf("%d pushed\n",val);
}
void pop(){
    if(top == NULL){
        printf("Stack underflow\n");
        return;
    }
    struct Node* temp = top;
    printf("%d popped\n",top->data);
    top = top->next;
    free(temp);
}

int peek(){
    if(top == NULL){
        printf("Stack is empty\n");
        return -1;
    }
    return top->data;
}

void display(){
    if (top == NULL){
        printf("Stack is empty\n");
        return;
    }
    struct Node* temp = top;
    printf("Stack ELements: ");
    while(temp != NULL){
        printf("%d ",temp->data);
        temp = temp->next;
    }
    printf("\n");
}

int main() {
    push(10);
    push(20);
    push(30);
    display();
    printf("Top element: %d\n", peek());
    pop();
    display();
    return 0;
}

File: Practise/stack_using_LL.py
class Node:
    def __init__(self,data):
        self.data = data
        self.next = None

class Stack:
    def __init__(self):
        self.top = None

    def push(self,val):
        new_node = Node(val)
        new_node.next = self.top
        self.top = new_node
        print(f"{val} pushed")

    def pop(self):
        if self.top is None:
            print("Stack underflow\n")
            return
        val = self.top.data
        self.top = self.top.next
        print(f"{val} pooped")

    def peek(self):
        if self.top is None:
            print("Stack is empty")
            return None
        return self.top.data

    def display(self):
        if self.top is None:
            print("Stack is empty")
            return
        temp = self.top
        print("Stack elements:", end=" ")
        while temp:
            print(temp.data, end=" ")
            temp = temp.next
        print()

s = Stack()
s.push(10)
s.push(20)
s.push(30)
s.display()
print("Top element:", s.peek())
s.pop()
s.display()